safespace/mgs/prefect_phenoscore.py

```diff
@@ -0,0 +1,800 @@
+    -------
+    Bunch
+        data bunch suitable for using in transmogrifier which contains metadata
+         and features dataframes, with matching non-hierarchical indices.  The
+         metadata contains an experiment column, and this Bunch is both 
+         aggregated to the site level and combined across experiments.
+    """    
+    data_dict = {exp: create_data(exp, modify_perts_func=modify_perts_func,
+                                  prescriptive=prescriptive)
+                 for exp in exps}
+    data_dict = {exp: clean_metadata(data, fill_dict=FILL_DICT)
+                 for exp, data in data_dict.items()}
+
+    # Aggregate patches to site level
+    group_by = ['experiment', 'plate_number', 'address', 'site']
+    grouped_dict = {exp: tm.pandas.groupby_aggregate(data=data,
+                                                     by=group_by,
+                                                     features_aggregations='mean')
+                    for exp, data in data_dict.items()}
+    combined = tm.pandas.concat(grouped_dict.values())
+    return combined
+
+
+# Transform
+
+@task
+def transform_features(combined, empty_query):
+    """transforms the features in a data bunch, applying plate center scaling, 
+    PCA transform, and plate center scaling.  The center scaling is applied to
+    a subset of the wells in the empty query (those that are used for scaling
+    are labeled with origin_group = 1)
+
+    Parameters
+    ----------
+    combined : Data Bunch
+        Data Bunch (can be of multiple experiments) that contains metadata and 
+        features
+    empty_query : string suitable for use as a pandas.DataFrame query
+        the query to get the candidates for the origin group.  Half of this 
+        group is used for center scaling, marked with origin_group = 1
+
+    Returns
+    -------
+    Bunch
+        transformed features and metadata, with added origin_group column
+    """    
+    # Set up reference query to use in analysis later
+    # empty_query = "well_type == 'MCSF-alone'"
+    combined['metadata']['origin_group'] = 0
+    to_center_scale = (combined['metadata'].query(empty_query).groupby('plate_number', as_index=False)
+                       .apply(lambda x: x.sample(frac=.5, replace=False)))
+    to_center_scale = to_center_scale.reset_index(drop=True, level=0)
+    combined['metadata'].loc[to_center_scale.index, 'origin_group'] = 1
+    origin_query = empty_query + " and origin_group == 1"
+    pipe = tm.pipeline.Pipeline([
+        ('ss', tm.preprocessing.StandardScaler(
+            metadata_query=origin_query, groupby_args=('plate_number',))),
+        ('pca', tm.decomposition.PCA(whiten=False)),
+        ('prs', tm.preprocessing.StandardScaler(
+            metadata_query=origin_query, groupby_args=('plate_number',)))
+    ])
+    combined_transformed = pipe.fit_transform(combined)
+    return combined_transformed
+
+
+# Z-factor calculations
+@task
+def calculate_zfactor(relabeled_data,
```

Comment:
```suggestion
def calculate_zfactor(*, relabeled_data,
```
This function has a lot of very specific named keywords, which would be easy to accidentally pass in the wrong order. I would use a `*` to require passing by name not position.

Reply:
Also not sure if it makes sense to give defaults to some of these? Or are they too custom/specific for each use case?